```
/Users/Shared/jenkins/workspace/OSX-64/xbmc/addons/test/TestAddonInfoBuilder.cpp:132:11: error: no member named 'SetType' in 'ADDON::CAddonInfoBuilder::CFromDB'
07:04:28   builder.SetType(ADDON_PLUGIN);
07:04:28   ~~~~~~~ ^
07:04:28 1 error generated.
07:04:28 make[3]: *** [CMakeFiles/kodi-test.dir/xbmc/addons/test/TestAddonInfoBuilder.cpp.o] Error 1
07:04:28 make[3]: *** Waiting for unfinished jobs....
07:04:28 /Users/Shared/jenkins/workspace/OSX-64/xbmc/addons/test/TestAddonDatabase.cpp:41:85: error: no viable conversion from 'vector<shared_ptr<ADDON::IAddon>>' to 'const vector<shared_ptr<ADDON::CAddonInfo>>'
07:04:28     database.UpdateRepositoryContent("repository.a", AddonVersion("1.0.0"), "test", addons);
07:04:28                                                                                     ^~~~~~
07:04:28 /Applications/Xcode10.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/vector:551:5: note: candidate constructor not viable: no known conversion from 'ADDON::VECADDONS' (aka 'vector<shared_ptr<ADDON::IAddon> >') to 'const std::__1::vector<std::__1::shared_ptr<ADDON::CAddonInfo>, std::__1::allocator<std::__1::shared_ptr<ADDON::CAddonInfo> > > &' for 1st argument
07:04:28     vector(const vector& __x);
07:04:28     ^
07:04:28 /Applications/Xcode10.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/vector:558:5: note: candidate constructor not viable: no known conversion from 'ADDON::VECADDONS' (aka 'vector<shared_ptr<ADDON::IAddon> >') to 'initializer_list<std::__1::vector<std::__1::shared_ptr<ADDON::CAddonInfo>, std::__1::allocator<std::__1::shared_ptr<ADDON::CAddonInfo> > >::value_type>' (aka 'initializer_list<std::__1::shared_ptr<ADDON::CAddonInfo> >') for 1st argument
07:04:28     vector(initializer_list<value_type> __il);
```

--- a/xbmc/addons/AddonDatabase.cpp
+++ b/xbmc/addons/AddonDatabase.cpp
@@ -26,7 +26,7 @@
 
 using namespace ADDON;
 
-static std::string SerializeMetadata(const CAddonInfo& addon)
+std::string CAddonDatabaseSerializer::SerializeMetadata(const CAddonInfo& addon)
 {
   CVariant variant;
   variant["author"] = addon.Author();
@@ -47,7 +47,7 @@ static std::string SerializeMetadata(const CAddonInfo& addon)
     variant["screenshots"].push_back(item);
 
   variant["extensions"] = CVariant(CVariant::VariantTypeArray);
-  variant["extensions"].push_back(ADDON::CAddonInfo::TranslateType(addon.MainType(), false));
+  variant["extensions"].push_back(SerializeExtensions(*addon.Type(addon.MainType())));
 
   variant["dependencies"] = CVariant(CVariant::VariantTypeArray);
   for (const auto& dep : addon.GetDependencies())
@@ -74,7 +74,42 @@ static std::string SerializeMetadata(const CAddonInfo& addon)
   return json;
 }
 
-static void DeserializeMetadata(const std::string& document, CAddonInfoBuilder::CFromDB& builder)
+CVariant CAddonDatabaseSerializer::SerializeExtensions(const CAddonExtensions& addonType)
+{
+  CVariant variant;
+  variant["type"] = addonType.m_point;
+
+  variant["values"] = CVariant(CVariant::VariantTypeArray);
+  for (const auto& value : addonType.m_values)
+  {
+    CVariant info(CVariant::VariantTypeObject);
+    info["id"] = value.first;
+    info["content"] = CVariant(CVariant::VariantTypeArray);
+    for (const auto& content : value.second)
+    {
+      CVariant contentEntry(CVariant::VariantTypeObject);
+      contentEntry["key"] = content.first;
+      contentEntry["value"] = content.second.str;
+      info["content"].push_back(std::move(contentEntry));
+    }
+
+    variant["values"].push_back(std::move(info));
+  }
+
+  variant["children"] = CVariant(CVariant::VariantTypeArray);
+  for (auto& child : addonType.m_children)
+  {
+    CVariant info(CVariant::VariantTypeObject);
+    info["id"] = child.first;
+    info["child"] = SerializeExtensions(child.second);
+    variant["children"].push_back(std::move(info));
+  }
+
+  return variant;
+}
+
+void CAddonDatabaseSerializer::DeserializeMetadata(const std::string& document,
+                                                   CAddonInfoBuilder::CFromDB& builder)
 {
   CVariant variant;
   if (!CJSONVariantParser::Parse(document, variant))
@@ -103,7 +138,10 @@ static void DeserializeMetadata(const std::string& document, CAddonInfoBuilder::
     screenshots.push_back(it->asString());
   builder.SetScreenshots(std::move(screenshots));
 
-  builder.SetType(CAddonInfo::TranslateType(variant["extensions"][0].asString()));
+  CAddonType addonType;
+  DeserializeExtensions(variant["extensions"][0], addonType);
+  addonType.m_type = CAddonInfo::TranslateType(addonType.m_point);
+  builder.SetExtensions(std::move(addonType));
 
   {
     std::vector<DependencyInfo> deps;
@@ -119,6 +157,37 @@ static void DeserializeMetadata(const std::string& document, CAddonInfoBuilder::
   for (auto it = variant["extrainfo"].begin_array(); it != variant["extrainfo"].end_array(); ++it)
     extraInfo.emplace((*it)["key"].asString(), (*it)["value"].asString());
   builder.SetExtrainfo(std::move(extraInfo));
+}
+
+void CAddonDatabaseSerializer::DeserializeExtensions(const CVariant& variant,
+                                                     CAddonExtensions& addonType)
+{
+  addonType.m_point = variant["type"].asString();
+
+  for (auto value = variant["values"].begin_array(); value != variant["values"].end_array();
+       ++value)
+  {
+    std::string id = (*value)["id"].asString();
+    std::vector<std::pair<std::string, SExtValue>> extValues;
+    for (auto content = (*value)["content"].begin_array();
+         content != (*value)["content"].end_array(); ++content)
+    {
+      extValues.emplace_back((*content)["key"].asString(), (*content)["value"].asString());
+    }
+
+    addonType.m_values.emplace_back(id, extValues);
+  }
+
+  for (auto child = variant["children"].begin_array(); child != variant["children"].end_array();
+       ++child)
+  {
+    CAddonExtensions childExt;
+    DeserializeExtensions((*child)["child"], childExt);
+    std::string id = (*child)["id"].asString();
+    addonType.m_children.emplace_back(id, childExt);
+  }
+
+  return;
 }
 
 CAddonDatabase::CAddonDatabase() = default;
@@ -435,7 +504,7 @@ bool CAddonDatabase::FindByAddonId(const std::string& addonId, ADDON::VECADDONS&
       builder.SetName(m_pDS->fv("name").get_asString());
       builder.SetSummary(m_pDS->fv("summary").get_asString());
       builder.SetDescription(m_pDS->fv("description").get_asString());
-      DeserializeMetadata(m_pDS->fv("metadata").get_asString(), builder);
+      CAddonDatabaseSerializer::DeserializeMetadata(m_pDS->fv("metadata").get_asString(), builder);
       builder.SetChangelog(m_pDS->fv("news").get_asString());
       builder.SetOrigin(m_pDS->fv("repoID").get_asString());
 
@@ -582,7 +651,7 @@ bool CAddonDatabase::GetAddon(int id, AddonPtr &addon)
     builder.SetName(m_pDS2->fv("name").get_asString());
     builder.SetSummary(m_pDS2->fv("summary").get_asString());
     builder.SetDescription(m_pDS2->fv("description").get_asString());
-    DeserializeMetadata(m_pDS2->fv("metadata").get_asString(), builder);
+    CAddonDatabaseSerializer::DeserializeMetadata(m_pDS2->fv("metadata").get_asString(), builder);
 
     addon = CAddonBuilder::Generate(builder.get(), ADDON_UNKNOWN);
     return addon != nullptr;
@@ -666,7 +735,7 @@ bool CAddonDatabase::GetRepositoryContent(const std::string& id, VECADDONS& addo
       builder.SetSummary(m_pDS->fv("summary").get_asString());
       builder.SetDescription(m_pDS->fv("description").get_asString());
       builder.SetOrigin(m_pDS->fv("repoID").get_asString());
-      DeserializeMetadata(m_pDS->fv("metadata").get_asString(), builder);
+      CAddonDatabaseSerializer::DeserializeMetadata(m_pDS->fv("metadata").get_asString(), builder);
 
       auto addon = CAddonBuilder::Generate(builder.get(), ADDON_UNKNOWN);
       if (addon)
@@ -774,13 +843,9 @@ bool CAddonDatabase::UpdateRepositoryContent(const std::string& repository,
       m_pDS->exec(PrepareSQL(
           "INSERT INTO addons (id, metadata, addonID, version, name, summary, description, news) "
           "VALUES (NULL, '%s', '%s', '%s', '%s','%s', '%s','%s')",
-          SerializeMetadata(*addon).c_str(),
-          addon->ID().c_str(),
-          addon->Version().asString().c_str(),
-          addon->Name().c_str(),
-          addon->Summary().c_str(),
-          addon->Description().c_str(),
-          addon->ChangeLog().c_str()));
+          CAddonDatabaseSerializer::SerializeMetadata(*addon).c_str(), addon->ID().c_str(),
+          addon->Version().asString().c_str(), addon->Name().c_str(), addon->Summary().c_str(),
+          addon->Description().c_str(), addon->ChangeLog().c_str()));
 
       int idAddon = static_cast<int>(m_pDS->lastinsertid());
       if (idAddon <= 0)

--- a/xbmc/addons/AddonDatabase.cpp
+++ b/xbmc/addons/AddonDatabase.cpp
@@ -26,7 +26,7 @@
 
 using namespace ADDON;
 
-static std::string SerializeMetadata(const IAddon& addon)
+static std::string SerializeMetadata(const CAddonInfo& addon)
 {
   CVariant variant;
   variant["author"] = addon.Author();
@@ -47,7 +47,7 @@ static std::string SerializeMetadata(const IAddon& addon)
     variant["screenshots"].push_back(item);
 
   variant["extensions"] = CVariant(CVariant::VariantTypeArray);
-  variant["extensions"].push_back(ADDON::CAddonInfo::TranslateType(addon.Type(), false));
+  variant["extensions"].push_back(ADDON::CAddonInfo::TranslateType(addon.MainType(), false));
 
   variant["dependencies"] = CVariant(CVariant::VariantTypeArray);
   for (const auto& dep : addon.GetDependencies())
@@ -747,8 +747,10 @@ int CAddonDatabase::GetRepositoryId(const std::string& addonId)
   return m_pDS->fv("id").get_asInt();
 }
 
-bool CAddonDatabase::UpdateRepositoryContent(const std::string& repository, const AddonVersion& version,
-    const std::string& checksum, const std::vector<AddonPtr>& addons)
+bool CAddonDatabase::UpdateRepositoryContent(const std::string& repository,
+                                             const AddonVersion& version,
+                                             const std::string& checksum,
+                                             const std::vector<AddonInfoPtr>& addons)
 {
   try
   {

--- a/xbmc/addons/AddonDatabase.cpp
+++ b/xbmc/addons/AddonDatabase.cpp
@@ -206,7 +206,7 @@ int CAddonDatabase::GetMinSchemaVersion() const
 
 int CAddonDatabase::GetSchemaVersion() const
 {
-  return 32;
+  return 33;
 }
 
 void CAddonDatabase::CreateTables()
@@ -317,6 +317,51 @@ void CAddonDatabase::UpdateTables(int version)
                 "blacklist");
     m_pDS->exec("DROP INDEX IF EXISTS idxBlack");
     m_pDS->exec("DROP TABLE blacklist");
+  }
+  if (version < 33)
+  {
+    m_pDS->query(PrepareSQL("SELECT * FROM addons"));
+    while (!m_pDS->eof())
+    {
+      const int id = m_pDS->fv("id").get_asInt();
+      const std::string metadata = m_pDS->fv("metadata").get_asString();
+
+      CVariant variant;
+      if (!CJSONVariantParser::Parse(metadata, variant))
+        continue;
+
+      CVariant variantUpdate;
+      variantUpdate["type"] = variant["extensions"][0].asString();
+      variantUpdate["values"] = CVariant(CVariant::VariantTypeArray);
+      variantUpdate["children"] = CVariant(CVariant::VariantTypeArray);
+
+      for (auto it = variant["extrainfo"].begin_array(); it != variant["extrainfo"].end_array();
+           ++it)
+      {
+        if ((*it)["key"].asString() == "provides")
+        {
+          CVariant info(CVariant::VariantTypeObject);
+          info["id"] = (*it)["key"].asString();
+          info["content"] = CVariant(CVariant::VariantTypeArray);
+
+          CVariant contentEntry(CVariant::VariantTypeObject);
+          contentEntry["key"] = (*it)["key"].asString();
+          contentEntry["value"] = (*it)["value"].asString();
+          info["content"].push_back(std::move(contentEntry));
+
+          variantUpdate["values"].push_back(std::move(info));
+          break;
+        }
+      }
+      variant["extensions"][0] = variantUpdate;
+
+      std::string json;
+      CJSONVariantWriter::Write(variant, json, true);
+      m_pDS->exec(PrepareSQL("UPDATE addons SET metadata='%s' WHERE id=%i", json.c_str(), id));
+
+      m_pDS->next();
+    }
+    m_pDS->close();
   }
 }
 

--- a/xbmc/addons/AddonDatabase.h
+++ b/xbmc/addons/AddonDatabase.h
@@ -11,6 +11,7 @@
 #include "AddonBuilder.h"
 #include "FileItem.h"
 #include "addons/Addon.h"
+#include "addons/addoninfo/AddonInfoBuilder.h"
 #include "dbwrappers/Database.h"
 
 #include <string>
@@ -18,6 +19,28 @@
 
 namespace ADDON
 {
+
+/*!
+ * @brief Addon content serializer/deserializer.
+ *
+ * Used to save data from the add-on in the database using json format.
+ * The corresponding field in SQL is "addons" for "metadata".
+ *
+ * @warning Changes in the json format need a way to update the addon database
+ * for users, otherwise problems may occur when reading the old content.
+ */
+class CAddonDatabaseSerializer
+{
+  CAddonDatabaseSerializer() = delete;
+
+public:
+  static std::string SerializeMetadata(const CAddonInfo& addon);
+  static void DeserializeMetadata(const std::string& document, CAddonInfoBuilder::CFromDB& builder);
+
+private:
+  static CVariant SerializeExtensions(const CAddonExtensions& addonType);
+  static void DeserializeExtensions(const CVariant& document, CAddonExtensions& addonType);
+};
 
 class CAddonDatabase : public CDatabase
 {

--- a/xbmc/addons/AddonDatabase.h
+++ b/xbmc/addons/AddonDatabase.h
@@ -45,8 +45,10 @@ public:
   /*! Returns all addons in the repositories with id `addonId`. */
   bool FindByAddonId(const std::string& addonId, ADDON::VECADDONS& addons) const;
 
-  bool UpdateRepositoryContent(const std::string& repositoryId, const ADDON::AddonVersion& version,
-      const std::string& checksum, const std::vector<ADDON::AddonPtr>& addons);
+  bool UpdateRepositoryContent(const std::string& repositoryId,
+                               const ADDON::AddonVersion& version,
+                               const std::string& checksum,
+                               const std::vector<AddonInfoPtr>& addons);
 
   int GetRepoChecksum(const std::string& id, std::string& checksum);
 

--- a/xbmc/addons/AddonManager.cpp
+++ b/xbmc/addons/AddonManager.cpp
@@ -897,36 +897,6 @@ bool CAddonMgr::IsAutoUpdateable(const std::string& id) const
   return m_updateRules.IsAutoUpdateable(id);
 }
 
-bool CAddonMgr::AddonsFromRepoXML(const CRepository::DirInfo& repo, const std::string& xml, VECADDONS& addons)
-{
-  CXBMCTinyXML doc;
-  if (!doc.Parse(xml))
-  {
-    CLog::Log(LOGERROR, "CAddonMgr::{}: Failed to parse addons.xml", __FUNCTION__);
-    return false;
-  }
-
-  if (doc.RootElement() == nullptr || doc.RootElement()->ValueStr() != "addons")
-  {
-    CLog::Log(LOGERROR, "CAddonMgr::{}: Failed to parse addons.xml. Malformed", __FUNCTION__);
-    return false;
-  }
-
-  // each addon XML should have a UTF-8 declaration
-  auto element = doc.RootElement()->FirstChildElement("addon");
-  while (element)
-  {
-    auto addonInfo = CAddonInfoBuilder::Generate(element, repo);
-    auto addon = CAddonBuilder::Generate(addonInfo, ADDON_UNKNOWN);
-    if (addon)
-      addons.push_back(std::move(addon));
-
-    element = element->NextSiblingElement("addon");
-  }
-
-  return true;
-}
-
 bool CAddonMgr::IsCompatible(const IAddon& addon) const
 {
   for (const auto& dependency : addon.GetDependencies())
@@ -1137,6 +1107,37 @@ bool CAddonMgr::SetAddonOrigin(const std::string& addonId, const std::string& re
   const AddonInfoPtr info = GetAddonInfo(addonId);
   if (info)
     m_database.GetInstallData(info);
+  return true;
+}
+
+bool CAddonMgr::AddonsFromRepoXML(const CRepository::DirInfo& repo,
+                                  const std::string& xml,
+                                  std::vector<AddonInfoPtr>& addons)
+{
+  CXBMCTinyXML doc;
+  if (!doc.Parse(xml))
+  {
+    CLog::Log(LOGERROR, "CAddonMgr::{}: Failed to parse addons.xml", __func__);
+    return false;
+  }
+
+  if (doc.RootElement() == nullptr || doc.RootElement()->ValueStr() != "addons")
+  {
+    CLog::Log(LOGERROR, "CAddonMgr::{}: Failed to parse addons.xml. Malformed", __func__);
+    return false;
+  }
+
+  // each addon XML should have a UTF-8 declaration
+  auto element = doc.RootElement()->FirstChildElement("addon");
+  while (element)
+  {
+    auto addonInfo = CAddonInfoBuilder::Generate(element, repo);
+    if (addonInfo)
+      addons.emplace_back(addonInfo);
+
+    element = element->NextSiblingElement("addon");
+  }
+
   return true;
 }
 

--- a/xbmc/addons/AddonManager.h
+++ b/xbmc/addons/AddonManager.h
@@ -304,15 +304,6 @@ namespace ADDON
      */
     bool LoadAddonDescription(const std::string &path, AddonPtr &addon);
 
-    /*! \brief Parse a repository XML file for addons and load their descriptors
-     A repository XML is essentially a concatenated list of addon descriptors.
-     \param repo The repository info.
-     \param xml The XML document from repository.
-     \param addons [out] returned list of addons.
-     \return true if the repository XML file is parsed, false otherwise.
-     */
-    bool AddonsFromRepoXML(const CRepository::DirInfo& repo, const std::string& xml, VECADDONS& addons);
-
     bool ServicesHasStarted() const;
 
     /*!
@@ -409,9 +400,6 @@ namespace ADDON
      *
      * @warning This should be never used from other places outside of addon
      * system directory.
-     *
-     * Currently listed call sources:
-     * - @ref CAddonInstallJob::DoWork
      */
     /*@{{{*/
 
@@ -429,8 +417,28 @@ namespace ADDON
      * @param[in] isUpdate If call becomes done on already installed addon and
      *                     update only.
      * @return True if successfully done, otherwise false
+     *
+     * Currently listed call sources:
+     * - @ref CAddonInstallJob::DoWork
      */
     bool SetAddonOrigin(const std::string& addonId, const std::string& repoAddonId, bool isUpdate);
+
+    /*!
+     * @brief Parse a repository XML file for addons and load their descriptors.
+     *
+     * A repository XML is essentially a concatenated list of addon descriptors.
+     *
+     * @param[in] repo The repository info.
+     * @param[in] xml The XML document from repository.
+     * @param[out] addons returned list of addons.
+     * @return true if the repository XML file is parsed, false otherwise.
+     *
+     * Currently listed call sources:
+     * - @ref CRepository::FetchIndex
+     */
+    bool AddonsFromRepoXML(const CRepository::DirInfo& repo,
+                           const std::string& xml,
+                           std::vector<AddonInfoPtr>& addons);
 
     /*@}}}*/
 

--- a/xbmc/addons/Repository.cpp
+++ b/xbmc/addons/Repository.cpp
@@ -183,7 +183,9 @@ bool CRepository::FetchChecksum(const std::string& url,
   return true;
 }
 
-bool CRepository::FetchIndex(const DirInfo& repo, std::string const& digest, VECADDONS& addons) noexcept
+bool CRepository::FetchIndex(const DirInfo& repo,
+                             std::string const& digest,
+                             std::vector<AddonInfoPtr>& addons) noexcept
 {
   XFILE::CCurlFile http;
 
@@ -222,7 +224,7 @@ bool CRepository::FetchIndex(const DirInfo& repo, std::string const& digest, VEC
 
 CRepository::FetchStatus CRepository::FetchIfChanged(const std::string& oldChecksum,
                                                      std::string& checksum,
-                                                     VECADDONS& addons,
+                                                     std::vector<AddonInfoPtr>& addons,
                                                      int& recheckAfter) const
 {
   checksum = "";
@@ -260,7 +262,7 @@ CRepository::FetchStatus CRepository::FetchIfChanged(const std::string& oldCheck
 
   for (const auto& dirTuple : dirChecksums)
   {
-    VECADDONS tmp;
+    std::vector<AddonInfoPtr> tmp;
     if (!FetchIndex(std::get<0>(dirTuple), std::get<1>(dirTuple), tmp))
       return STATUS_ERROR;
     addons.insert(addons.end(), tmp.begin(), tmp.end());
@@ -324,7 +326,7 @@ bool CRepositoryUpdateJob::DoWork()
     oldChecksum = "";
 
   std::string newChecksum;
-  VECADDONS addons;
+  std::vector<AddonInfoPtr> addons;
   int recheckAfter;
   auto status = m_repo->FetchIfChanged(oldChecksum, newChecksum, addons, recheckAfter);
 

--- a/xbmc/addons/Repository.h
+++ b/xbmc/addons/Repository.h
@@ -46,7 +46,7 @@ namespace ADDON
 
     FetchStatus FetchIfChanged(const std::string& oldChecksum,
                                std::string& checksum,
-                               VECADDONS& addons,
+                               std::vector<AddonInfoPtr>& addons,
                                int& recheckAfter) const;
 
     struct ResolveResult
@@ -60,7 +60,9 @@ namespace ADDON
     static bool FetchChecksum(const std::string& url,
                               std::string& checksum,
                               int& recheckAfter) noexcept;
-    static bool FetchIndex(const DirInfo& repo, std::string const& digest, VECADDONS& addons) noexcept;
+    static bool FetchIndex(const DirInfo& repo,
+                           std::string const& digest,
+                           std::vector<AddonInfoPtr>& addons) noexcept;
 
     static DirInfo ParseDirConfiguration(const CAddonExtensions& configuration);
 

--- a/xbmc/addons/addoninfo/AddonExtensions.h
+++ b/xbmc/addons/addoninfo/AddonExtensions.h
@@ -18,6 +18,7 @@ namespace ADDON
 {
 
 class CAddonInfoBuilder;
+class CAddonDatabaseSerializer;
 
 struct SExtValue
 {
@@ -67,6 +68,7 @@ public:
 
 private:
   friend class CAddonInfoBuilder;
+  friend class CAddonDatabaseSerializer;
 
   std::string m_point;
   EXT_VALUES m_values;

--- a/xbmc/addons/addoninfo/AddonInfoBuilder.h
+++ b/xbmc/addons/addoninfo/AddonInfoBuilder.h
@@ -57,21 +57,19 @@ public:
     void SetExtrainfo(InfoMap extrainfo)
     {
       m_addonInfo->m_extrainfo = std::move(extrainfo);
-
-      const auto& it = m_addonInfo->m_extrainfo.find("provides");
-      if (it != m_addonInfo->m_extrainfo.end())
-      {
-        CAddonType addonType(m_addonInfo->m_mainType);
-        addonType.SetProvides(it->second);
-        m_addonInfo->m_types.push_back(addonType);
-      }
     }
-    void SetType(TYPE type) { m_addonInfo->m_mainType = type; }
     void SetInstallDate(const CDateTime& installDate) { m_addonInfo->m_installDate = installDate; }
     void SetLastUpdated(const CDateTime& lastUpdated) { m_addonInfo->m_lastUpdated = lastUpdated; }
     void SetLastUsed(const CDateTime& lastUsed) { m_addonInfo->m_lastUsed = lastUsed; }
     void SetOrigin(std::string origin) { m_addonInfo->m_origin = std::move(origin); }
     void SetPackageSize(uint64_t size) { m_addonInfo->m_packageSize = size; }
+    void SetExtensions(CAddonType addonType)
+    {
+      if (!addonType.GetValue("provides").empty())
+        addonType.SetProvides(addonType.GetValue("provides").asString());
+      m_addonInfo->m_types.push_back(std::move(addonType));
+      m_addonInfo->m_mainType = addonType.m_type;
+    }
 
     const AddonInfoPtr& get() { return m_addonInfo; }
 

--- a/xbmc/addons/addoninfo/AddonType.h
+++ b/xbmc/addons/addoninfo/AddonType.h
@@ -72,6 +72,7 @@ typedef enum
 } TYPE;
 
 class CAddonInfoBuilder;
+class CAddonDatabaseSerializer;
 
 class CAddonType : public CAddonExtensions
 {
@@ -108,6 +109,7 @@ public:
 
 private:
   friend class CAddonInfoBuilder;
+  friend class CAddonDatabaseSerializer;
 
   void SetProvides(const std::string& content);
 

--- a/xbmc/addons/test/TestAddonBuilder.cpp
+++ b/xbmc/addons/test/TestAddonBuilder.cpp
@@ -34,7 +34,8 @@ TEST_F(TestAddonBuilder, ShouldBuildDependencyAddons)
   CAddonInfoBuilder::CFromDB builder;
   builder.SetId("aa");
   builder.SetDependencies(deps);
-  builder.SetType(ADDON_UNKNOWN);
+  CAddonType addonType(ADDON_UNKNOWN);
+  builder.SetExtensions(addonType);
   AddonPtr addon = CAddonBuilder::Generate(builder.get(), ADDON_UNKNOWN);
   EXPECT_EQ(deps, addon->GetDependencies());
 }
@@ -43,7 +44,8 @@ TEST_F(TestAddonBuilder, ShouldReturnDerivedType)
 {
   CAddonInfoBuilder::CFromDB builder;
   builder.SetId("aa");
-  builder.SetType(ADDON_RESOURCE_LANGUAGE);
+  CAddonType addonType(ADDON_RESOURCE_LANGUAGE);
+  builder.SetExtensions(addonType);
   auto addon = std::dynamic_pointer_cast<CLanguageResource>(CAddonBuilder::Generate(builder.get(), ADDON_UNKNOWN));
   EXPECT_NE(nullptr, addon);
 }

--- a/xbmc/addons/test/TestAddonDatabase.cpp
+++ b/xbmc/addons/test/TestAddonDatabase.cpp
@@ -35,7 +35,7 @@ protected:
     std::set<std::string> installed{"repository.a", "repository.b"};
     database.SyncInstalled(installed, installed, std::set<std::string>());
 
-    VECADDONS addons;
+    std::vector<AddonInfoPtr> addons;
     CreateAddon(addons, "foo.bar", "1.0.0");
     database.SetRepoUpdateData("repository.a", {});
     database.UpdateRepositoryContent("repository.a", AddonVersion("1.0.0"), "test", addons);
@@ -46,13 +46,12 @@ protected:
     database.UpdateRepositoryContent("repository.b", AddonVersion("1.0.0"), "test", addons);
   }
 
-  void CreateAddon(VECADDONS& addons, std::string id, std::string version)
+  void CreateAddon(std::vector<AddonInfoPtr>& addons, std::string id, std::string version)
   {
     CAddonInfoBuilder::CFromDB builder;
     builder.SetId(id);
     builder.SetVersion(AddonVersion(version));
-    AddonPtr addon = CAddonBuilder::Generate(builder.get(), ADDON_UNKNOWN);
-    addons.push_back(addon);
+    addons.push_back(builder.get());
   }
 
   void TearDown() override

--- a/xbmc/addons/test/TestAddonInfoBuilder.cpp
+++ b/xbmc/addons/test/TestAddonInfoBuilder.cpp
@@ -129,7 +129,9 @@ TEST_F(TestAddonInfoBuilder, TestGenerate_DBEntry)
   CAddonInfoBuilder::CFromDB builder;
   builder.SetId("video.blablabla.org");
   builder.SetVersion(AddonVersion("1.2.3"));
-  builder.SetType(ADDON_PLUGIN);
+  CAddonType addonType(ADDON_PLUGIN);
+  addonType.Insert("provides", "video audio");
+  builder.SetExtensions(addonType);
   builder.SetName("The Bla Bla Bla Addon");
   builder.SetAuthor("Team Kodi");
   builder.SetSummary("Summary bla bla bla");
@@ -141,7 +143,6 @@ TEST_F(TestAddonInfoBuilder, TestGenerate_DBEntry)
   builder.SetEMail("a@a.dummy");
   builder.SetSource("https://github.com/xbmc/xbmc");
   InfoMap extrainfo;
-  extrainfo["provides"] = "video audio";
   extrainfo["language"] = "marsian";
   builder.SetExtrainfo(extrainfo);
 

